Assets/LeapMotion/Modules/InteractionEngine/Scripts/InteractionManager.cs

```diff
@@ -830,13 +854,93 @@ public class InteractionManager : MonoBehaviour, IRuntimeGizmoComponent {
         bool shouldIgnore = Physics.GetIgnoreLayerCollision(_templateLayer, i);
         Physics.IgnoreLayerCollision(_interactionLayer, i, shouldIgnore);
         Physics.IgnoreLayerCollision(_interactionNoContactLayer, i, shouldIgnore);
-
-        // Set brush layer to collide with nothing
-        Physics.IgnoreLayerCollision(_contactBoneLayer, i, true);
+        Physics.IgnoreLayerCollision(_contactBoneLayer, i, shouldIgnore);
       }
 
-      //After copy and set we enable the interaction between the brushes and interaction objects
+      // Enable interactions between the contact bones and the interaction layer.
       Physics.IgnoreLayerCollision(_contactBoneLayer, _interactionLayer, false);
+
+      // Disable interactions between the contact bones and the no-contact layer.
+      Physics.IgnoreLayerCollision(_contactBoneLayer, _interactionNoContactLayer, true);
+    }
+
+    #endregion
+
+    #region Interaction Object Layer Tracking
+
+    private Dictionary<SingleLayer, HashSet<IInteractionBehaviour>> _intObjInteractionLayers = new Dictionary<SingleLayer, HashSet<IInteractionBehaviour>>();
+    private Dictionary<SingleLayer, HashSet<IInteractionBehaviour>> _intObjNoContactLayers   = new Dictionary<SingleLayer, HashSet<IInteractionBehaviour>>();
+
+    private int _interactionLayerMask = 0;
+    /// <summary>
+    /// Returns a layer mask containing all layers that might contain interaction objects.
+    /// </summary>
+    public int GetInteractionLayerMask() {
+      return _interactionLayerMask;
+    }
+
+    private void refreshInteractionLayerMask() {
+      _interactionLayerMask = 0;
+
+      // Accumulate single-layer layer masks into the combined interaction layer mask.
+      foreach (var layerObjSetPair in _intObjInteractionLayers) {
+        _interactionLayerMask = layerObjSetPair.Key.layerMask | _interactionLayerMask;
+      }
+      foreach (var layerObjSetPair in _intObjNoContactLayers) {
+        _interactionLayerMask = layerObjSetPair.Key.layerMask | _interactionLayerMask;
+      }
+    }
+
+    void IInternalInteractionManager.RefreshLayersNow() {
+      refreshInteractionLayerMask();
+    }
+
+    void IInternalInteractionManager.NotifyIntObjAddedInteractionLayer(IInteractionBehaviour intObj, int layer, bool refreshImmediately) {
+      if (!_intObjInteractionLayers.ContainsKey(layer)) {
+        _intObjInteractionLayers[layer] = new HashSet<IInteractionBehaviour>();
+      }
+
+      _intObjInteractionLayers[layer].Add(intObj);
+
+      if (refreshImmediately) {
+        refreshInteractionLayerMask();
+      }
+    }
+
+    void IInternalInteractionManager.NotifyIntObjRemovedInteractionLayer(IInteractionBehaviour intObj, int layer, bool refreshImmediately) {
+      _intObjInteractionLayers[layer].Remove(intObj);
+
+      if (_intObjInteractionLayers[layer].Count == 0) {
+        _intObjInteractionLayers.Remove(layer);
```

Comment:
Instead of removing the hash set, why not leave it?  You can always check if it is empty in the refresh method for very little extra cost, and that way you are able to avoid the case where a specific object with a unique override layer is being enabled/disabled a whole bunch, causing a new hash set to be constructed every time it is enabled.

It's also better to keep hash sets around.  As objects are added to them, they need to periodically resize, but they keep their size once they have expanded.  If you delete the hash set when it becomes empty, it needs to go through that resizing process again, resulting in extra allocations and time.